[master] Improve salt-ssh error handling

lkubb opened the pull request:
~~Note: This will stay a draft until https://github.com/saltstack/salt/pull/64576 finds its way into master (or the regression is fixed otherwise)~~ ~~Since there has been no activity on the referenced PR for months, I went ahead and merged it into this branch to unblock this PR.~~ It's in. :)

### What does this PR do?
* Ensures that state rendering is interrupted when an execution module run on the target exits with a non-zero exit code or returns non-decodable or invalid data. Thereby ensures that templates are not rendered with garbage inputs. Wrapper modules will receive an exception and thus will not mistake error returns for remote function output. See https://github.com/saltstack/salt/issues/52452#issuecomment-1601065486 for details.
* Excludes error returns from mine data, as it is done in usual Salt as well.
* Makes `state.*` wrappers treat non-decodable json output (usually caused by Salt crashing completely and printing a stacktrace only) as an error condition.
* Ensures salt-ssh job ret events contain all usual data (saw it while reading the code and included the minor fix here, hope that's OK)

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/52452
Fixes: https://github.com/saltstack/salt/issues/64531
Fixes: https://github.com/saltstack/salt/issues/42065
Fixes: https://github.com/saltstack/salt/issues/50727 (last-minute addition: after re-reading the report, I think it's relatively clear why it was not working before and that this should fix it, see test in [tests/pytests/integration/ssh/state/test_retcode_state_run_remote_exception.py](https://github.com/saltstack/salt/pull/64542/files#diff-ae71f89f956935339d33b0b39496d0c77d51b532f65075d89b42d7feec716467))

### Previous Behavior
* Executing any non-wrapped module inside a template or wrapper module would imply the possibility that errors are evaluated as truthy or randomly rendered salt-ssh error return dicts would mess up the YAML etc. structure when printing the output.
* Mine function returns could contain the same garbage.
* Fatal errors on the remote during state execution would be treated as OK.

### New Behavior
* Exceptions are raised when a remote run returns with a non-zero exit code, returns data that indicates failure or if JSON output was expected, but deserialization failed. This causes template rendering to abort and wrapper modules to be confronted with an exception.
* Error returns are excluded from the mine.
* Fatal errors on the remote during state execution return a non-zero exit code.

### Merge requirements satisfied?
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

lkubb committed:
Add tests for issues #64531 and #52452

lkubb committed:
Make salt-ssh more strict when handling unexpected situations

This commit ensures that output indicating failure is not
passed off as the output of the command when it is called
via `__salt__`, e.g. during template rendering or in wrapper
modules. It does this by raising exceptions when issues
are detected and thus simulates the remote error locally.
See issue #64531 for a detailed description.

lkubb committed:
Adapt remote execution error output tests

lkubb committed:
Make `state.*` wrappers treat a remote exception as failure

lkubb committed:
Exclude salt-ssh error returns from mine

lkubb committed:
Run pyupgrade

lkubb committed:
Filter out deprecation warnings from reported stderr

lkubb committed:
Also test exact expected command parsing failure output

lkubb committed:
Update key deploy routine for new parsing workflow

lkubb committed:
Ensure salt-ssh job ret events contain all usual data

lkubb committed:
Cleanup, add changelog entry for compound issue

lkubb committed:
Fix SSHCommandExecutionError string casting output

It should include the specific error, not just the general one

lkubb committed:
Also test exact expected state failure output

Ch3LL reviewed (changes requested):
I haven't reviewed this entirely yet, but this seems like it should go into the `master` branch due to the new classes and functions that are being added. You okay with that approach?

lkubb commented:
@Ch3LL Yup, was unsure whether to submit this against master or 3006.x. I will wait with the rebase onto master until the state fixes from Friday are merged forward, which this expands upon.

lkubb committed:
Adapt unit tests, report permission denied as before

Also fix an integration test which was broken by a botched
rebase conflict resolution.

lkubb committed:
Use `find_json` again, more validation

lkubb commented:
@Ch3LL This should be ready for review again, I merged the retcode fix into this branch since it has been in limbo for months. The changes to `client.ssh.shell` are therefore unrelated to this PR.

lkubb committed:
Fix new unit tests

s0undt3ch commented on tests/pytests/integration/ssh/test_state.py:
```diff
@@ -548,16 +572,174 @@ class TestStateRunFailRetcode:
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def test_retcode_state_sls_id_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.sls_id", "foo", "fail_module_exception")
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def test_retcode_state_show_sls_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.show_sls", "fail_module_exception")
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def test_retcode_state_show_low_sls_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.show_low_sls", "fail_module_exception")
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def test_retcode_state_show_highstate_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.show_highstate")
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def test_retcode_state_show_lowstate_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.show_lowstate")
+        # state.show_lowstate exits with 0 for non-ssh as well
+        self._assert_ret(ret, 0)
+
+    def test_retcode_state_top_render_module_exception(self, salt_ssh_cli):
+        ret = salt_ssh_cli.run("state.top", "top.sls")
+        self._assert_ret(ret, EX_AGGREGATE)
+
+    def _assert_ret(self, ret, retcode):
+        assert ret.returncode == retcode
+        assert isinstance(ret.data, list)
+        assert ret.data
+        assert isinstance(ret.data[0], str)
+        # While these three should usually follow each other, there
+        # can be warnings in between that would break such a logic.
+        assert "Rendering SLS 'base:fail_module_exception' failed:" in ret.data[0]
+        assert "Problem running salt function in Jinja template:" in ret.data[0]
+        assert (
+            "Error running 'disk.usage': Invalid flag passed to disk.usage"
+            in ret.data[0]
+        )
```
```suggestion
@pytest.mark.usefixtures("state_tree_render_module_exception")
@pytest.mark.parametrize(
    "args,retcode",
    (
        (("state.sls", "fail_module_exception"), EX_AGGREGATE),
        (("state.highstate",), EX_AGGREGATE),
        (("state.sls_id", "foo", "fail_module_exception"), EX_AGGREGATE),
        (("state.show_sls", "fail_module_exception"), EX_AGGREGATE),
        (("state.show_low_sls", "fail_module_exception"), EX_AGGREGATE),
        (("state.show_highstate",), EX_AGGREGATE),
        # state.show_lowstate exits with 0 for non-ssh as well
        (("state.show_lowstate",), 0),
        (("state.top", "top.sls"), EX_AGGREGATE),
    ),
)
def test_retcode_render_module_exception(salt_ssh_cli, args, retcode):
    """
    Verify salt-ssh stops state execution and fails with a retcode > 0
    when a state rendering fails because an execution module throws an exception.
    """
    ret = salt_ssh_cli.run(*args)
    
    assert ret.returncode == retcode
    assert isinstance(ret.data, list)
    assert ret.data
    assert isinstance(ret.data[0], str)
    # While these three should usually follow each other, there
    # can be warnings in between that would break such a logic.
    assert "Rendering SLS 'base:fail_module_exception' failed:" in ret.data[0]
    assert "Problem running salt function in Jinja template:" in ret.data[0]
    assert (
        "Error running 'disk.usage': Invalid flag passed to disk.usage"
        in ret.data[0]
    )
```

lkubb replied:
The intention was similar as with the other classes - to save resources by making the fixture class-scoped. I will refactor this in a the same way as with the ssh pillar fixes.

s0undt3ch replied:
Why can't it be `module` scoped then?

lkubb replied:
Refactored the state tests in the same style as https://github.com/saltstack/salt/pull/65484 combining both of your suggestions (class -> module with single parametrized test function).

I didn't want to name them twice (same module name and test name), so shortened the test name.

s0undt3ch commented on tests/pytests/integration/ssh/test_state.py:
```diff
@@ -548,16 +572,174 @@ class TestStateRunFailRetcode:
+
+@pytest.mark.slow_test
+@pytest.mark.usefixtures("state_tree_remote_exception")
+class TestStateRunRemoteExceptionRetcode:
```
This can be a single, parametrized, test, like shown above.

s0undt3ch commented on tests/pytests/unit/client/ssh/wrapper/test_parse_ret.py:
```diff
@@ -0,0 +1,46 @@
+    stderr = "Permission denied, please try again.\nPermission denied, please try again.\nroot@192.168.1.187: Permission denied (publickey,gssapi-keyex,gssapi-with-micimport pudb; pu.dbassword).\nscp: Connection closed\n"
+    retcode = 255
+
+    try:
+        wrap.parse_ret(stdout, stderr, retcode)
+    except wrap.SSHPermissionDeniedError as err:
+        # need access to the exception instance, which pytest.raises
+        # does not provide
+        ret = err.to_ret()
+    else:
+        assert False, "Did not raise SSHPermissionDeniedError"
+    assert "_error" in ret
+    assert ret["_error"] == "Permission denied"
+    assert "stdout" in ret
+    assert "stderr" in ret
+    assert "Permission denied (publickey" in ret["stderr"]
+    assert "retcode" in ret
+    assert ret["retcode"] == 255
```
```suggestion
    with pytest.raises(wrap.SSHPermissionDeniedError) as exc:
        wrap.parse_ret(stdout, stderr, retcode)
    ret = err.value.to_ret()
    assert "_error" in ret
    assert ret["_error"] == "Permission denied"
    assert "stdout" in ret
    assert "stderr" in ret
    assert "Permission denied (publickey" in ret["stderr"]
    assert "retcode" in ret
    assert ret["retcode"] == 255
```
This doesn't work?

s0undt3ch commented on tests/pytests/unit/client/ssh/wrapper/test_parse_ret.py:
```diff
@@ -0,0 +1,46 @@
+    try:
+        wrap.parse_ret(stdout, stderr, retcode)
+    except wrap.SSHPermissionDeniedError:
+        assert False, "This should not have resulted in an SSHPermissionDeniedError"
```
```suggestion
        pytest.fail("This should not have resulted in an SSHPermissionDeniedError")
```

lkubb committed:
Apply suggested changes by review

s0undt3ch commented on salt/client/ssh/__init__.py:
```diff
@@ -464,19 +464,15 @@ def key_deploy(self, host, ret):
             target = self.targets[host]
             if target.get("passwd", False) or self.opts["ssh_passwd"]:
                 self._key_deploy_run(host, target, False)
-            return ret
```
@Ch3LL need your review on this bit, I know you worked on it recently

lkubb replied:
@Ch3LL just making sure you saw Pedro's request. :)

lkubb commented:
@s0undt3ch Anything we can do to get this into 3007? This represents a crucial fix to Salt-SSH reliability imo, otherwise every single unwrapped call has to wrapped with a return value check (or surprising behavior has to be expected).

dwoz commented:
@lkubb please rebase.

lkubb commented:
@dwoz Rebased.

The multiple merges with conflicts made this rebase pretty challenging, but I checked it against the diff of the previous branch with master merged and everything looks fine.

dwoz commented:
@lkubb Awesome, lets hope all the tests pass now :D
